### PR TITLE
Fixed compilation with MinGW and did some cleanup

### PR DESCRIPTION
**Pull request by drfrag666**

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -51,28 +51,11 @@ if( WIN32 )
 	add_definitions( -D_WIN32 )
 
 
-	if( ( MSVC14 AND NOT CMAKE_GENERATOR_TOOLSET STREQUAL "v140_xp" ) OR	# For VS 2015.
-		( MSVC15 AND NOT CMAKE_GENERATOR_TOOLSET STREQUAL "v141_xp" ) )		# For VS 2017.
+	if( MSVC AND MSVC_VERSION GREATER 1699 )	# For VS 2015 and later.
```
Comment (alexey-lysiuk) [inline src/CMakeLists.txt]: `MSVC_VERSION` is equal to 1700 [for VS2012](https://cmake.org/cmake/help/v3.16/variable/MSVC_VERSION.html). Moreover, we do not support VS2015 anymore.

**Comment by drfrag666** [comment]
I made a copy-pasta error. I checked this: https://cmake.org/cmake/help/latest/variable/MSVC_VERSION.html#variable:MSVC_VERSION
 I wanted to put 1899 there, the comment didn't mean that version was supported.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -51,28 +51,11 @@ if( WIN32 )
 	add_definitions( -D_WIN32 )
 
 
-	if( ( MSVC14 AND NOT CMAKE_GENERATOR_TOOLSET STREQUAL "v140_xp" ) OR	# For VS 2015.
-		( MSVC15 AND NOT CMAKE_GENERATOR_TOOLSET STREQUAL "v141_xp" ) )		# For VS 2017.
+	if( MSVC AND MSVC_VERSION GREATER 1900 )	# For VS 2017 and later.
```
Comment (alexey-lysiuk) [inline src/CMakeLists.txt]: What is the case you want to handle with this version check? In other words, `if( MSVC )` seems to be enough.

Comment (drfrag666) [reply src/CMakeLists.txt]: True, older versions are not actually supported. In my branch i support 2015.

**Comment by coelckers** [comment]
Will be merged once reviewed.

**Comment by drfrag666** [comment]
Changes not being related is a questionable matter. I don't think that XInput stuff is questionable. That stuff was required for older versions of VS, XP and MinGW. While i was at it i had to do the cleanup. Now that thing is not needed anymore (it's in the else and that part now is only for MinGW and possibly Clang on windows) and was used to compile without XInput. Then the XP stuff was not even needed for XP with VS 2015 and later.
${CMAKE_DL_LIBS} is required even for VS but has to be linked somewhere else master would not compile, the legacy branch doesn't so probably is something Vulkan related. I could leave that part out but i don't see why, i haven't tried to compile without Vulkan yet.
https://forum.zdoom.org/viewtopic.php?f=35&t=66800#p1129855
 But that said @Blzut3 added all that XP and Xinput stuff so he could have a look at this if that makes you feel safer.

**Comment by alexey-lysiuk** [comment]
> Changes not being related is a questionable matter. I don't think that XInput stuff is questionable. That stuff was required for older versions of VS, XP and MinGW. While i was at it i had to do the cleanup. Now that thing is not needed anymore (it's in the else and that part now is only for MinGW and possibly Clang on windows) and was used to compile without XInput. Then the XP stuff was not even needed for XP with VS 2015 and later.

OK, I would ask @Blzut3 to review these changes then.

> ${CMAKE_DL_LIBS} is required for MinGW and linux but has to be linked somewhere else master would not compile, the legacy branch doesn't so probably is something Vulkan related. I could leave that part out but i don't see why, i haven't tried to compile without Vulkan yet.
> https://forum.zdoom.org/viewtopic.php?f=35&t=66800#p1129855

> Edit: ${CMAKE_DL_LIBS} is not required for VS but i don't see any problem adding it, technically it should be required right?

No, `${CMAKE_DL_LIBS}` isn't needed for VS, and it's not defined for MSVC. In this context, unset variable is equal to an empty string, so it works without any `if()`'s.

**Comment by drfrag666** [comment]
I've already asked him.
Then i'll remove that ${CMAKE_DL_LIBS} for VS.

**Comment by alexey-lysiuk** [comment]
> Then i'll remove that ${CMAKE_DL_LIBS} for VS.

What for?

**Comment by drfrag666** [comment]
True, it would be just a cosmetic change.

**Comment by drfrag666** [comment]
Now i remember what that NO_XINPUT stuff was for, it was to compile with old versions of MinGW which lacked XInput support.

**Comment by Blzut3** [comment]
If it works as it should on Visual Studio with these changes then I don't see anything in the pull request to object to.  I'd maybe personally like to see the change in vk_mem_alloc.h to be on the same line as the other conditions.

But yes, the ability to build without XINPUT was for MinGW.  Without booting up Windows I believe w64 ships with XInput and since we're not supporting Windows 98 there's not much reason to look at non-w64 MinGW.  But if drfrag says that change came from me (the patch in question was from like 2016 or thereabouts) then that would mean TDM-GCC supported XInput and the ability to compile without it is pointless.

**Comment by drfrag666** [comment]
Well it has to work as those changes are in the else. About the XP checks for me it compiles and links with v141_xp even without those.
On that condition i considered that but IMO it was already rather complex.
Both MinGW-w64 versions come with XInput, TDM-GCC didn't compile at first but later i found it did with the include from the DX SDK. Wasn't that patch precisely for TDM-GCC? TDM-GCC can't compile GZDoom since long ago.

**Comment by Blzut3** [comment]
I guess I'm just surprised the checks for dxguid being MinGW exclusive were not needed.

Anyway by TDM-GCC I meant non-w64 TDM-GCC.  My patch way back was done specifically to make C++11 GZDoom work on Windows 98 to disprove the whole "no updated toolchain is available" thing (and it did work).  However the patch was geared towards using w64 out of the box and hacking non-w64's system headers.  Obviously TDM-GCC is irrelevant now, just providing context for the patch you referred to and why I might not be able to remember specifics.

**Comment by drfrag666** [comment]
Those checks were removed and i added them back for MinGW only but now since older VS versions are not supported anymore everything under the else is only for MinGW and may be others. They are still required.

**Comment by drfrag666** [comment]
I've rebased it but there wouldn't have been any conflict, i've deleted the CMakelists for ZMusic.
I still don't see how XInput could fail as i've only removed the NO_XINPUT part and in CMakelists the changes are in the non MSVC part (else) .

**Pull request merged**